env-var-docs/generate_markdown.py

```diff
@@ -0,0 +1,155 @@
+import os
+import sys
+import typing
+
+
+def errorexit(msg):
+    """Logs a message and exits"""
+    sys.stderr.write(msg + "\n")
+    exit(1)
+
+
+@dataclasses.dataclass
+class Config:
+    docs_path: str
+    version: str
+    local_output: bool
+    access_token: str
+    repo: str
+    repo_path: str
+
+
+def make_config() -> Config:
+    """Returns a Config by reading variables set in the environment. See docs
+    at the beginning of this file for required and optional variables.
+
+    Exits if any there are any validation errors.
+    """
+    if "ENV_VAR_DOCS_PATH" not in os.environ:
+        errorexit(
+            "ENV_VAR_DOCS_PATH must be present in the environment variables")
+
+    docs = os.environ["ENV_VAR_DOCS_PATH"]
+    if not os.path.isfile(docs):
+        errorexit(f"'{docs}' does not point to a file")
+
+    local = (os.environ.get("LOCAL_OUTPUT", "false") == "true")
+    if not local:
+        try:
+            version = os.environ["RELEASE_VERSION"]
+            token = os.environ["GITHUB_ACCESS_TOKEN"]
+            repo = os.environ["TARGET_REPO"]
+            path = os.environ["TARGET_PATH"]
+            if path[0:1] == "/":
+                errorexit("f{path} must be a relative path")
+        except KeyError as e:
+            errorexit(
+                f"Either LOCAL_OUTPUT=true must be set or {e.args[0]} must be present in the environment variables"
+            )
+    else:
+        version = token = repo = path = ""
+
+    return Config(docs, version, local, token, repo, path)
+
+
+def main():
+    config = make_config()
+    with open(config.docs_path) as docs_file:
+        markdown, parse_errors = generate_markdown(docs_file)
+        if len(parse_errors) != 0:
+            msg = f"{config.docs_path} is invalid:\n"
+            msg += env_var_docs.errors_formatter.format(parse_errors)
+            errorexit(msg)
+
+    if config.local_output:
+        print(markdown)
+        exit()
+
+    gh = github.Github(config.access_token)
+    repo = gh.get_repo(config.repo)
+
+    # If file exists, update it, if not, create it.
+    path = f"{config.repo_path}/{config.version}.md"
+    msg = f"Adds server environment variable documentation for {config.version}"
+    try:
+        file = repo.get_contents(path)
+        if isinstance(file, list):
+            errorexit(
+                f"{file_path} returns multiple files in the repo, aborting")
+
+        res = repo.update_file(path, msg, markdown, file.sha)
+    except github.GithubException.UnknownObjectException:
+        res = repo.create_file(path, msg, markdown, branch="main")
+
+    print(
+        f"https://github.com/blob/main/{config.repo}/{path} updated in commit {res['commit']}"
+    )
+
+
+def generate_markdown(
+    docs_file: typing.TextIO
+) -> tuple[str | None, list[env_var_docs.parser.NodeParseError]]:
+    out = ""
+
+    def output(node: env_var_docs.parser.Node):
+        nonlocal out  # Need to declare out nonlocal otherwise it gets shadowed.
```

Comment:
`out` is the markdown document that we are building up.  Because `output` (renamed to `append_node_to_out`) is a callback passed to the parsing package `visit` function, it needs to capture `out` so that it can write to it.  We need to declare `out` as nonlocal for the capture to work